src/callback.h

```diff
@@ -0,0 +1,248 @@
+        }
+        if (reduced) {
+            reduce(*mulf);
+        }
+    }
+
+    virtual void OnIteration(int type, void *data, uint64_t iteration) = 0;
+
+    form* forms;
+    int64_t iterations = 0;
+    integer D;
+    integer L;
+    PulmarkReducer* reducer;
+    vdf_original* vdfo;
+};
+
+class OneWesolowskiCallback: public WesolowskiCallback {
+  public:
+    OneWesolowskiCallback(integer& D, uint64_t wanted_iter) : WesolowskiCallback(D) {
+        int k, l;
+        this->wanted_iter = wanted_iter;
+        ApproximateParameters(wanted_iter, k, l);
+        kl = k * l;
+        uint64_t space_needed = wanted_iter / (k * l) + 100;
+        forms = (form*) calloc(space_needed, sizeof(form));
+        form f = form::generator(D);
+        forms[0] = f;
+    }
+
+    ~OneWesolowskiCallback() {
+        delete(vdfo);
+        delete(reducer);
+        free(forms);
+    }
+
+    void OnIteration(int type, void *data, uint64_t iteration) {
+        iteration++;
+        if (iteration > wanted_iter) 
+            return ;
+
+        if (iteration % kl == 0) {
+            uint64_t pos = iteration / kl;
+            form* mulf = &forms[pos];
+            SetForm(type, data, mulf);
+        }
+        if (iteration == wanted_iter) {
+            SetForm(type, data, &result);
+        }
+    }
+
+    uint64_t wanted_iter;
+    uint64_t kl;
+    form result;
+};
+
+class TwoWesolowskiCallback: public WesolowskiCallback {
+  public:
+    TwoWesolowskiCallback(integer& D) : WesolowskiCallback(D) {
+        int space_needed = kSwitchIters / 10 + (kMaxItersAllowed - kSwitchIters) / 100;
+        forms = (form*) calloc(space_needed, sizeof(form));
+        form f = form::generator(D);
+        forms[0] = f;
+        kl = 10;
+        switch_iters = -1;
+    }
+
+    ~TwoWesolowskiCallback() {
+        delete(vdfo);
+        delete(reducer);
+        free(forms);
+    }
+
+    void IncreaseConstants(uint64_t num_iters) {
+        kl = 100;
+        switch_iters = num_iters;
+        switch_index = num_iters / 10;
+    }
+
+    int GetPosition(uint64_t power) {
+        if (switch_iters == -1 || power < switch_iters) {
+            return power / 10;
+        } else {
+            return (switch_index + (power - switch_iters) / 100);
+        }
+    }
+
+    form *GetForm(uint64_t power) {
+        return &(forms[GetPosition(power)]);
+    }
+
+    bool LargeConstants() {
+        return kl == 100;
+    }
+
+    void OnIteration(int type, void *data, uint64_t iteration) {
+        iteration++;
+        if (iteration % kl == 0) {
+            uint64_t pos = GetPosition(iteration);
+            form* mulf = &forms[pos];
+            SetForm(type, data, mulf);
+        }
+    }
+
+  private:
+    uint64_t switch_index;
+    int64_t switch_iters;
+    uint32_t kl;
+};
+
+class FastAlgorithmCallback : public WesolowskiCallback {
+  public:
+    FastAlgorithmCallback(int segments, integer& D, bool fast_machine) : WesolowskiCallback(D) {
```

Comment:
I think we should call this flag/variable multi_proc_machine - which seems to be what you're using it for